src/python/pants/option/global_options.py

```diff
@@ -81,14 +81,25 @@ class ExecutionOptions:
     remote_execution_process_cache_namespace: Any
     remote_instance_name: Any
     remote_ca_certs_path: Any
-    remote_oauth_bearer_token_path: Any
     remote_execution_extra_platform_properties: Any
-    remote_execution_headers: Any
+    remote_execution_headers: Union[None, List[Tuple[str, str]], Callable[[Optional[str]], dict]]
     remote_execution_overall_deadline_secs: int
     process_execution_local_enable_nailgun: bool
 
     @classmethod
     def from_bootstrap_options(cls, bootstrap_options):
+        headers = bootstrap_options.remote_execution_headers
```

Comment:
This is definitely not the final form of how to parse the options. The goal of this code was to demonstrate that the supplying a closure works.

The main question here is how to plugins would register themselves for this hook.